### ISSUE TITLE
sparse=True still materializes the dense Jacobian and Hessian to detect the pattern

### ISSUE THREAD

**Issue by jkitchin** [jkitchin/pounce#464]
# `sparse=True` still materializes the dense Jacobian and Hessian to detect the pattern

**Version:** pounce 0.9.0, editable install from a `main` working tree at
`a0c7776`. Python 3.13.5, macOS.

## Summary

`from_jax(..., sparse=True)` compresses the per-iteration Jacobian and Hessian
*evaluation* with CPR coloring, which works well. But the sparsity **detection**
that precedes it evaluates the dense `(m, n)` Jacobian and the dense `(n, n)`
Lagrangian Hessian at `n_probes` random points, unconditionally — before the
`if sparse:` branch is ever reached.

So the build is O(n²) in memory no matter what `sparse` is set to, and on a
genuinely sparse problem that ceiling is hit long before the solve would be.

## Where

`python/pounce/jax/_build.py`, `_JaxProblem.__init__` (~line 122):

```python
# --- detect sparsity from a union of random probes ---
rng = np.random.default_rng(seed)
x_probes = [jnp.asarray(rng.standard_normal(n)) for _ in range(n_probes)]
if m > 0:
    jac_denses = [_to_np(self._jac_g_dense(xp)) for xp in x_probes]   # (m, n) dense
    self._jac_rows, self._jac_cols = _detect_pattern_2d_multi(jac_denses)
    hess_denses = [                                                    # (n, n) dense
        _to_np(self._hess_lag_dense(xp, lp, 1.0))
        for xp, lp in zip(x_probes, lam_probes)
    ]
...
# --- compressed (colored) AD callables, when requested ---
if sparse:
    self._build_compressed()
```

`n_probes` defaults to **3** when `sparse=True`, so the sparse path allocates
*more* dense matrices than the dense path, not fewer.

The torch backend looks the same: `python/pounce/torch/_problem.py` ~609-625
builds `jac_denses` / `hess_denses` the same way.

## Measurement

A Radau IIA collocation system from a transient gas-network study (block-banded
Jacobian, ~0.5% dense; `n = m`), built with `from_jax(..., sparse=True)`:

| `n_var` | build time |
|---|---|
| 2 268 | 2.6 s |
| 4 536 | 7.6 s |
| 9 072 | 30.1 s |
| 18 144 | killed at 10 min, 5.5 GB resident |

Clean 4x per doubling — the O(n²) signature. Note the *solve* over the same
range is 1.6 → 3.5 → 6.7 s, i.e. roughly linear and entirely reasonable. It is
only the setup that blows up.

At the size I actually wanted (`n_var = 91 584`) the probe alone is a
91584 × 91584 float64 matrix: 67 TB. The problem is not large by collocation
standards — it is 96 elements of a 159-state ODE — and its sparsity is known in
closed form.

## Why this is more than a performance nit

For a full-discretization/simultaneous method, knowing the Jacobian structure
analytically *is the premise of the approach*: element `i` couples only to
element `i-1`, so the pattern is block-banded before any numbers exist. Having
to rediscover that by dense probing inverts the method's main structural
advantage, and it means a benchmark of "simultaneous vs sequential" measures
the probe rather than the method.

There is also a correctness edge: detection is probabilistic (a structural zero
that coincides with the probe values is silently dropped), and `n_probes`
mitigates that only statistically. A caller who *knows* the pattern currently
has no way to say so.

## Requests, in order of value to me

1. **Let the caller supply the pattern.** Something like
   `from_jax(..., jac_pattern=(rows, cols), hess_pattern=(rows, cols))`,
   skipping detection entirely. This is the one that unblocks me, it is
   backward compatible, and it removes the probabilistic-detection risk for
   callers who have the structure in hand.
2. **Detect without materializing.** Under `sparse=True`, probe with a handful
   of seeded JVPs (and HVPs) and union the nonzero rows, which is O(n) memory
   per probe rather than O(n²). Still probabilistic, but it removes the ceiling.
3. **Failing both, document it.** The `sparse` docstring currently reads as
   though it makes the whole path cheap ("drops the per-eval cost from O(n) to
   O(k) AD passes"); it would help to say that build memory is O(n²) regardless.

## Workaround I am using

I build the pattern from the collocation structure directly, then hand pounce a
cyipopt-shaped problem object that reuses `pounce._ad_common._color_columns` and
`pounce.jax._build._seed_matrix` — i.e. exactly your compressed path, with the
pattern supplied instead of detected. That is ~150 lines sitting in my study's
`common/sparsity.py`, and it is entirely machinery that belongs in pounce.

Happy to turn it into a PR for request 1 if the API shape above looks right.
